modules/compatibility/src/test/java/org/apache/ignite/compatibility/sql/randomsql/RandomQuerySupplier.java

```diff
@@ -80,151 +92,211 @@ public RandomQuerySupplier(Schema schema, int seed) {
+     * @return Ast representing query filter.
      */
-    private Ast rndFrom(RandomisedQueryContext rndQryCtx) {
-        List<TableRef> tbls = rndTables(rndQryCtx.schema());
+    private Ast rndQueryFilter(RandomisedQueryContext rndQryCtx) {
+        List<TableRef> scopeTbls = rndQryCtx.scopeTables();
+
+        Ast cond1 = rndTableFilter(rndQryCtx, scopeTbls.get(rnd.nextInt(scopeTbls.size())));
 
-        assert !F.isEmpty(tbls);
+        if (dice.roll() <= 2)
+            return cond1;
 
-        if (tbls.size() == 1 || rndWithRatio(5)) {
-            tbls.forEach(rndQryCtx::addScopeTable);
+        Ast cond2 = rndTableFilter(rndQryCtx, scopeTbls.get(rnd.nextInt(scopeTbls.size())));
 
-            return new TableList(tbls);
+        return new BiCondition(cond1, cond2, dice.roll() <= 2 ? Operator.AND : Operator.OR);
+    }
+
+    /**
+     * Generates random filter for specified table.
+     *
+     * @param rndQryCtx Context of randomised query.
+     * @param tbl Table that requires filter.
+     * @return Ast representing table filter.
+     */
+    private Ast rndTableFilter(RandomisedQueryContext rndQryCtx, TableRef tbl) {
+        Ast rigthOp;
+        if (dice.roll() == 1)
+            rigthOp = new Const(Integer.toString(rnd.nextInt(100)));
+        else {
+            rigthOp = new Const("?");
+
+            rndQryCtx.addQueryParam(rnd.nextInt(100));
         }
 
-        TableRef leftTbl = tbls.get(0);
+        return new BiCondition(
+            rndNumericColumn(tbl),
+            rigthOp,
+            Operator.EQUALS
+        );
+    }
 
-        Ast from = leftTbl;
+    /**
+     * Picks N random table from the schema and adds them to the scope.
+     *
+     * @param rndQryCtx Random query context.
+     * @param tblCnt Tables count.
+     */
+    private void pickRndTables(RandomisedQueryContext rndQryCtx, int tblCnt) {
+        pickRndTables(rndQryCtx, tblCnt, false);
+    }
 
-        rndQryCtx.addScopeTable(leftTbl);
+    /**
+     * Picks N random table from the schema and adds them to the scope.
+     *
+     * @param rndQryCtx Random query context.
+     * @param tblCnt Tables count.
+     * @param allowDublicates Whether to allow dublicates or not.
+     */
+    private void pickRndTables(RandomisedQueryContext rndQryCtx, int tblCnt, boolean allowDublicates) {
+        List<Table> tbls = new ArrayList<>(schema.tables());
 
-        for (int i = 1; i < tbls.size(); i++) {
-            TableRef rightTbl = tbls.get(i);
+        if (allowDublicates)
+            tbls.addAll(schema.tables()); // add all tables again to get duplicates
 
-            from = new InnerJoin(from, rightTbl,
-                new BiCondition(rndNumericColumn(rndScopeTable(rndQryCtx)), rndNumericColumn(rightTbl), Operator.EQUALS));
+        Collections.shuffle(tbls, rnd);
 
-            rndQryCtx.addScopeTable(rightTbl); // add a table to the scope AFTER we created the join condition
-        }
+        for (Table tbl : tbls.subList(0, tblCnt)) {
```

Comment:
can we require more than 6 tables (with duplicates) from schema with only 3 table?

Reply:
yes, we can. So I've add handling for this case, thx